source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1027,6 +1028,192 @@ ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks(ClusterUpdateCallbacks&
   return std::make_unique<ClusterUpdateCallbacksHandleImpl>(cb, cluster_manager.update_callbacks_);
 }
 
+namespace {
+
+using ClusterAddedCb = std::function<void(ThreadLocalCluster&)>;
+
+class ClusterCallbacks : public ClusterUpdateCallbacks {
+public:
+  ClusterCallbacks(ClusterAddedCb cb) : cb_(std::move(cb)) {}
+
+  void onClusterAddOrUpdate(ThreadLocalCluster& cluster) override { cb_(cluster); };
+
+  void onClusterRemoval(const std::string&) override {}
+
+private:
+  ClusterAddedCb cb_;
+};
+
+} // namespace
+
+ClusterManagerImpl::ClusterDiscoveryManager::ClusterDiscoveryManager(
+    ThreadLocalClusterManagerImpl& parent)
+    : parent_(parent) {}
+
+void ClusterManagerImpl::ClusterDiscoveryManager::ensureLifecycleCallbacksAreInstalled() {
```

Comment:
Why is this function needed? Can't these callbacks just be installed as part of the CDM constructor? Is the issue calling a virtual function in the constructor? For that I would slightly refactor to fix that vs. doing it this way. Let me know if you need more details.

Reply:
Well, my thinking was that we wouldn't want these callbacks to be installed if they are not used. So the idea was to install them on first use and remove them if all the requests were handled.

Reply:
I added `ThreadLocalClusterManagerImpl::addClusterUpdateCallbacks` for adding the cluster lifecycle callback, and both `ClusterManagerImpl::addThreadLocalClusterUpdateCallbacks` and `ClusterDiscoveryManager` constructor use it.